dhall/src/Dhall/TypeCheck.hs

```diff
@@ -937,86 +939,98 @@ infer typer = loop
                 then return ()
                 else die (UnusedHandler diffT)
 
-            (my₀, _T₁') <- do
-                case mT₁ of
-                    Just _T₁ -> do
-                        return (Nothing, eval values _T₁)
-
-                    Nothing -> do
-                        case Dhall.Map.uncons yTs' of
-                            Nothing -> do
-                                die MissingMergeType
+            if Data.Set.null diffU
+                then return ()
+                else die (MissingHandler diffU)
 
-                            Just (y₀, _T', _) -> do
-                                _T₁' <- do
-                                    case Dhall.Map.lookup y₀ yUs' of
-                                        Nothing -> do
-                                            die (UnusedHandler diffU)
+            let match _y _T₀' Nothing =
+                    return _T₀'
 
-                                        Just Nothing -> do
-                                            return _T'
+                match y handler' (Just _A₁') =
+                    case Eval.toVHPi handler' of
+                        Just (x, _A₀', _T₀') -> do
+                            if Eval.conv values _A₀' _A₁'
+                                then do
+                                    let _T₁' = _T₀' (fresh ctx x)
 
-                                        Just (Just _) -> do
-                                            case Eval.toVHPi _T' of
-                                                Just (x, _A₀', _T₀') -> do
-                                                    return (_T₀' (fresh ctx x))
+                                    let _T₁'' = quote names _T₁'
 
-                                                Nothing -> do
-                                                    let _T'' = quote names _T'
+                                    -- x appearing in _T₁'' would indicate a dependently typed
```

Comment:
Thanks for the suggestions on the error message. Now I'm wondering whether I was mistaken to call the handler "dependently typed", and whether I should also update this comment.

Reply:
Yeah, I would just call it "disallowed" rather than "dependently typed"